snapshotter/devmapper/devmapper.go

```diff
@@ -0,0 +1,319 @@
+		return complete(ctx, trans, err)
+	}
+
+	return complete(ctx, trans, nil)
+}
+
+func (dm *Snapshotter) Remove(ctx context.Context, key string) error {
+	log.G(ctx).WithField("key", key).Debug("remove")
+
+	ctx, trans, err := dm.store.TransactionContext(ctx, true)
+	if err != nil {
+		return err
+	}
+
+	snapID, _, err := storage.Remove(ctx, key)
+	if err != nil {
+		return complete(ctx, trans, err)
+	}
+
+	deviceName := dm.getDeviceName(snapID)
+	if err := dm.pool.RemoveDevice(deviceName); err != nil {
+		log.G(ctx).WithError(err).Errorf("failed to remove device")
+		return complete(ctx, trans, err)
+	}
+
+	return complete(ctx, trans, nil)
+}
+
+func (dm *Snapshotter) Walk(ctx context.Context, fn func(context.Context, snapshots.Info) error) error {
+	log.G(ctx).Debug("walk")
+
+	ctx, trans, err := dm.store.TransactionContext(ctx, false)
+	if err != nil {
+		return err
+	}
+
+	defer trans.Rollback()
+	return storage.WalkInfo(ctx, fn)
+}
+
+func (dm *Snapshotter) Close() error {
+	log.L.Debug("close")
+
+	var result *multierror.Error
+
+	if err := dm.store.Close(); err != nil {
+		result = multierror.Append(result, err)
+	}
+
+	if err := dm.pool.Close(context.Background(), false, false); err != nil {
+		result = multierror.Append(result, err)
+	}
+
+	return result.ErrorOrNil()
+}
+
+func (dm *Snapshotter) createSnapshot(ctx context.Context, kind snapshots.Kind, key, parent string, opts ...snapshots.Opt) ([]mount.Mount, error) {
+	ctx, trans, err := dm.store.TransactionContext(ctx, true)
+	if err != nil {
+		return nil, err
+	}
+
+	snap, err := storage.CreateSnapshot(ctx, kind, key, parent, opts...)
+	if err != nil {
+		return nil, complete(ctx, trans, err)
+	}
+
+	if len(snap.ParentIDs) == 0 {
+		deviceName := dm.getDeviceName(snap.ID)
+		log.G(ctx).Debugf("creating new thin device '%s'", deviceName)
+
+		deviceID, err := dm.pool.CreateThinDevice(deviceName, dm.config.BaseImageSizeBytes)
+		if err != nil {
+			log.G(ctx).WithError(err).Errorf("failed to create thin device for snapshot %s", snap.ID)
+			return nil, complete(ctx, trans, err)
+		}
+
+		log.G(ctx).Debugf("created thin device with id %d", deviceID)
+		if err := dm.mkfs(ctx, deviceName); err != nil {
+			return nil, complete(ctx, trans, err)
+		}
+	} else {
+		parentDeviceName := dm.getDeviceName(snap.ParentIDs[0])
+		snapDeviceName := dm.getDeviceName(snap.ID)
+		log.G(ctx).Debugf("creating snapshot device '%s' from '%s'", snapDeviceName, parentDeviceName)
+
+		snapDeviceID, err := dm.pool.CreateSnapshotDevice(parentDeviceName, snapDeviceName, dm.config.BaseImageSizeBytes)
+		if err != nil {
+			log.G(ctx).WithError(err).Errorf("failed to create snapshot device from parent %s", parentDeviceName)
+			return nil, complete(ctx, trans, err)
+		}
+
+		log.G(ctx).Debugf("created snapshot device with id %d", snapDeviceID)
+	}
+
+	mounts := dm.buildMounts(snap)
+
+	// This is needed only to pass containerd's snapshotter suite tests
```

Comment:
Why does it fail if we exclude the `WithTempMount`?

Reply:
It doesn't. `lost+found` appears when writing file system with `mkfs` and it's not expected by containerd's snapshotter tests, and makes them fail. Though it's not needed for production.
I consider adding `deleteLostAndFound bool` and run this code only during tests.

Reply:
Test will be already failed and we don't have `mounts` (we don't know to what test `mounts` belongs to since we run a lot of them).

Reply:
How about splitting this function up a little bit then? Like `createSnapshot` returns the snapshot with no mounts binded. Then you have to call `bindMounts` to actually bind the mounts. You can then update the tests to handle removing of the special folders. Is this feasible?

Reply:
The part where "Then you have to call bindMounts to actually bind the mounts" is either called by containerd or containerd's snapshotter suite tests. We don't have control over it.

Reply:
Sorry not bind mount, but build mount. Basically you build the snapshot object then build the mounts in a separate function. This should give you all the information you need to remove the `lost+found`in the tests, correct?

Reply:
Have a look here https://github.com/firecracker-microvm/firecracker-containerd/pull/39/files#diff-a6d0ec2c2ea0600ed7470184982548a1R33
This code is called from an external test, which we can't modify.

Reply:
Ah, now I get what you are talking about. I wonder if it makes sense to refactor that test a little bit so we don't need to have that test code in there.

Reply:
This set of tests is used for all snapshotters in `containerd` project. I don't think it'll be refactored for our specific case.

Reply:
This one https://github.com/containerd/containerd/blob/master/snapshots/testsuite/testsuite.go#L190